LOOP-3973: Introducing `interruptionLevel` as a replacement for `Alert.isCritical`

rickpasetto opened the pull request:
This introduces the notion of three Alert "interruption levels" similar to iOS's
UNNotificationInterruptionLevel (https://developer.apple.com/documentation/usernotifications/unnotificationinterruptionlevel)
as a replacement of the `isCritical` flag.

https://tidepool.atlassian.net/browse/LOOP-3973

LWPR: https://github.com/tidepool-org/LoopWorkspace/pull/734

rickpasetto committed:
LOOP-3973: Introducing `interruptionLevel` as a replacement for `Alert.isCritical`

This introduces the notion of three Alert "interruption levels" similar to iOS's
UNNotificationInterruptionLevel (https://developer.apple.com/documentation/usernotifications/unnotificationinterruptionlevel)
as a replacement of the `isCritical` flag.

https://tidepool.atlassian.net/browse/LOOP-3973

rickpasetto committed:
Revert change accidentally pushed (that would make iOS critical alerts show up while app is in foreground)

rickpasetto committed:
LOOP-3974: Make Bluetooth Off a critical alert

darinkrauss commented on Loop/Managers/Alerts/AlertStore.xcdatamodeld/AlertStore.xcdatamodel/contents:
```diff
@@ -1,11 +1,11 @@
         <attribute name="backgroundContent" optional="YES" attributeType="String"/>
         <attribute name="foregroundContent" optional="YES" attributeType="String"/>
-        <attribute name="isCritical" attributeType="Boolean" usesScalarValueType="YES"/>
+        <attribute name="interruptionLevel" attributeType="Integer 16" usesScalarValueType="YES"/>
```
Update the the mappingmodel?

rickpasetto replied:
Ah, thanks for the reminder!

rickpasetto replied:
Ah wait, there is no mappingmodel, because this store doesn't exist out there in released form, so there's no need (yet) for a migration.

darinkrauss replied:
So it doesn't even need a placeholder in the mappingmodel file?

rickpasetto replied:
There is no mappingModel file.  I'd imagine we'd need to make one when we start migrating this object.

darinkrauss replied:
Should we version this and create one? I believe all of this is in DIY now, at least in the `dev` branch.

darinkrauss commented on Loop/Managers/Alerts/StoredAlert+CoreDataProperties.swift:
```diff
@@ -21,7 +21,7 @@ extension StoredAlert {
     @NSManaged public var backgroundContent: String?
     @NSManaged public var foregroundContent: String?
-    @NSManaged public var isCritical: Bool
+    @NSManaged public var interruptionLevel: Int16
```
Consider making this a primitive and have a getter/setter to nicely map to `Alert.InterruptionLevel`.

rickpasetto replied:
Actually, I decided to remove this altogether.  the (formerly) `isCritical` boolean was actually an aggregation of what is already stored in `foregroundContent` and `backgroundContent`.  As a `Bool` it sort of made sense to guess at it, but now that it is an enum I think it is duplicate at best and misleading at worst.

darinkrauss reviewed (changes requested):
I think you at least need to update the mappingmodel file.

rickpasetto committed:
Remove `interruptionLevel` from CoreData altogether.

It used to be a "convenience" to have the `isCritical` boolean there, but because `interruptionLevel` is really part of `Alert.Content` (and is serialized in the `foregroundContent` and `backgroundContent` fields already), I think keeping this field is  at best confusing and at worst misleading.

rickpasetto commented on Loop/Managers/Alerts/StoredAlert.swift:
```diff
@@ -22,7 +22,6 @@ extension StoredAlert {
             managerIdentifier = alert.identifier.managerIdentifier
             triggerType = alert.trigger.storedType
             triggerInterval = alert.trigger.storedInterval
-            isCritical = alert.foregroundContent?.isCritical ?? false || alert.backgroundContent?.isCritical ?? false
```
`interruptionLevel` is stored in the `foregroundContent` and `backgroundContent` fields.

rickpasetto committed:
Hoists interruptionLevel into Alert

rickpasetto committed:
Fix primitiveInterruptionLevel

rickpasetto committed:
Merge branch 'dev' into rick/LOOP-3973

darinkrauss commented on Loop/Managers/Alerts/StoredAlert+CoreDataClass.swift:
```diff
@@ -9,9 +9,23 @@
+        get {
+            willAccessValue(forKey: "interruptionLevel")
+            defer { didAccessValue(forKey: "interruptionLevel") }
+            return try! Alert.InterruptionLevel(storedValue: primitiveValue(forKey: "interruptionLevel") as! Int16)
```
The practice I've seen in other ManagedObjects is to have the primitive specified as `primitiveWhatever` and access that rather than using `primitiveValue(forKey:)` and `setPrimitiveValue(forKey:)`. See `CachedInsulinDeliveryObject` for an example.

rickpasetto replied:
I did that exact thing at first...but I kept getting mysterious `EXC_BAD_ACCESS` crashes immediately in the setter.  I did some research but could not find a root cause; I got as far as finding out it had something to do with the autogenerated code from `@NSManaged`. So I removed it and implemented it "by hand" using the `primitiveValue` and `setPrimitiveValue` primitives.  The docs really don't encourage this, but I couldn't find another way.  Any ideas?

darinkrauss replied:
What type did you use in the `StoredAlert+CoreDataProperties.swift` file? I believe it has to be `NSNumber`.

rickpasetto replied:
Ah, ok.  That's the problem, then.  I used Int16...and TBH now this is ringing a bell.  It all seems rather silly to use @NSManaged, but then have to write custom code for the primitive type, which can't really be primitive but an NSNumber.  🙄 

I'll switch it to NSNumber.

darinkrauss commented on Loop/Managers/Alerts/StoredAlert.swift:
```diff
@@ -16,13 +17,15 @@ extension StoredAlert {
     convenience init(from alert: Alert, context: NSManagedObjectContext, issuedDate: Date = Date()) {
         do {
-            self.init(context: context)
+            let name = String(describing: type(of: self))
```
What is this code for?

rickpasetto replied:
While running the unit tests, I kept seeing the following warnings:
```
CoreData: warning: Multiple NSEntityDescriptions claim the NSManagedObject subclass 'Loop.StoredAlert' so +entity is unable to disambiguate.
```
Note that these happened even before all of my changes.  This SO gives an example of how to mitigate this, which is what this is: https://stackoverflow.com/a/54126839

I'll add a comment explaining...

darinkrauss commented on Loop/Managers/Alerts/StoredAlert.swift:
```diff
@@ -145,6 +149,48 @@ extension Alert.Trigger {
+            if #available(iOS 15.0, *) {
+                return Int16(UNNotificationInterruptionLevel.active.rawValue)
+            } else {
+                return 1
```
[nit] Might want some indication that `UNNotificationInterruptionLevel.active.raw` **IS** `1`. My first thought was "is there a problem when upgrading from iOS 14 with a value of `1` to iOS 15 with a value of `Int16(UNNotificationInterruptionLevel.active.rawValue)`"?

rickpasetto replied:
I'll put a link to the Apple docs.

darinkrauss commented on Loop/Managers/Alerts/StoredAlert.swift:
```diff
@@ -145,6 +149,48 @@ extension Alert.Trigger {
+        }
+    }
+    
+    init(storedValue: Int16) throws {
```
A throw seems pretty heavy-handed here. The typical usage in Loop is to make it a failable initializer (`init?`).

rickpasetto replied:
heh, I went back and forth on this...I thought you'd like this better.  I guessed wrong :).  Will switch.

darinkrauss replied:
What I don't like is silently failing when failable initializers return `nil` during deserialization. If the key is found in the dictionary, but then deserialization of the value fails, then that is an error. Unfortunately, most of our deserialization code (typically via RawRepresentable), silently fails. 

Otherwise I'm fine with failable initializers and personally tend to steer away from throwing in the Loop code.

That's just my two cents, though...

rickpasetto replied:
My preference when decoding is to throw, because sometimes the type is used as an optional field.  Now we've got a situation: is the initialization failing due a coding error or is the field value actually nil?  So now all the callers (which deal with that field) need to check for nil, which means "coding error", which is not the same as the (optional) field just not being present.  

In other words, you can no longer do:

```
   field = optionalRawValue.map { MyEnum(from: $0) }
```

Because the `MyEnum` constructor might return nil.

Again, I also think this is all philosophical but in a system when I can influence the style guides, that's my preferred way to go.  Just my 2c.

darinkrauss replied:
I understand. Just note that this style isn't used in any of the rest of the Loop code (excepting the Decoder protocol conformance).

darinkrauss commented on Loop/Managers/LoopAlertsManager.swift:
```diff
@@ -50,7 +50,8 @@ public class LoopAlertsManager {
                                       acknowledgeActionButtonLabel: NSLocalizedString("Dismiss", comment: "Default alert dismissal"))
-        alertManager?.issueAlert(Alert(identifier: bluetoothPoweredOffIdentifier, foregroundContent: fgcontent, backgroundContent: bgcontent, trigger: .immediate))
+        alertManager?.issueAlert(Alert(identifier: bluetoothPoweredOffIdentifier, foregroundContent: fgcontent, backgroundContent: bgcontent, trigger: .immediate,
+                                       interruptionLevel: .critical))
```
[nit] Consider moving all to multiple lines, rather than just the last parameter. (Or collapse back to a single line.)

darinkrauss reviewed (approved):
Some things to consider changing, but otherwise okay.

rickpasetto committed:
Adds unit test to insure confidence in `interruptionLevel` getter/setter

rickpasetto committed:
PR Feedback
